Add change tracking methods for `belongs_to` associations

georgeclaghorn opened the pull request:
Permit checking whether a `belongs_to` association has been pointed to a new target record in the previous save and whether it will point to a new target record in the next save.

```ruby
post.category # => #<Category id: 1, name: "Ruby">

post.category = Category.second   # => #<Category id: 2, name: "Programming">
post.category_changed?            # => true
post.category_previously_changed? # => false

post.save!

post.category_changed?            # => false
post.category_previously_changed? # => true
```

Example use case: use Hotwire to broadcast removal of a blog post from its previous category when its category changes. Requires an accessor for the previous target of the association, which can be added separately.

To-do:

- [x] Documentation
- [x] Tests
- [x] Changelog entry

georgeclaghorn committed:
Add change tracking methods for belongs_to associations

Permit checking whether a belongs_to association has been pointed to a new target record in the previous save and whether it will point to a new target record in the next save.

    post.category # => #<Category id: 1, name: "Ruby">

    post.category = Category.second   # => #<Category id: 2, name: "Programming">
    post.category_changed?            # => true
    post.category_previously_changed? # => false

    post.save!

    post.category_changed?            # => false
    post.category_previously_changed? # => true

kaspth commented on activerecord/lib/active_record/associations/belongs_to_polymorphic_association.rb:
```diff
@@ -10,6 +10,14 @@ def klass
+      end
+
+      def target_previously_changed?
+        super || owner.attribute_changed?(reflection.foreign_type)
```
Should this use previously changed too?

georgeclaghorn replied:
Yep. Fixed.

eugeneius commented:
There's one more list of generated association methods that could be updated here:

https://github.com/rails/rails/blob/96fdf0d73749f721de27b9da41f88d57f95fc91c/activerecord/lib/active_record/associations.rb#L357-L366

eugeneius commented:
Seeing `belongs_to :polymorphic` in that table above made me realise that this class needs to be updated too:

https://github.com/rails/rails/blob/c2669b85ba9dfe99e783293b5c04c4c2121a4993/activerecord/lib/active_record/associations/belongs_to_polymorphic_association.rb#L12-L14

georgeclaghorn commented:
> There's one more list of generated association methods that could be updated here:

Thanks, took care of that one.

> Seeing `belongs_to :polymorphic` in that table above made me realise that this class needs to be updated too:

I think I had updated `BelongsToPolymorphicAssociation` already, but it needed test coverage, which I added [here](https://github.com/rails/rails/pull/42751/files#diff-6b40eaf3a990b978a06b57029195160ec17d81fe7a6c91ecfe388fc554f682e8R1589).

kaspth commented:
Thanks @georgeclaghorn! 🙌
